File: obsidian/Experiments/Group Theory/Burnside Group/B25/Infiniteness/data/stallings/fold.py
#!/usr/bin/env python3
"""Exact membership of a word in H_M = <d^5 : d reduced word over {a,b}, 1 <= |d| <= M>
via Stallings folding. H_M is an ordinary (non-normal) subgroup of F2; H_M increases to
the verbal subgroup F^5 as M -> infinity. So:  target in H_M  =>  target = 1 in B(2,5);
target not in H_M  =>  "not a product of fifth powers with root length <= M" (exact theorem).

Labels: 0=a,1=b,2=A,3=B ; inverse label = l^2 (xor 2).
Folding: worklist algorithm with union-find, adjacency = list of dicts.
"""
import sys
from collections import deque

INV = {0: 2, 1: 3, 2: 0, 3: 1}
C2L = {'a': 0, 'b': 1, 'A': 2, 'B': 3}

class Graph:
    def __init__(self):
        self.parent = []
        self.adj = []          # vertex -> {label: neighbor}
        self.base = self.new_vertex()

    def new_vertex(self):
        self.parent.append(len(self.parent))
        self.adj.append({})
        return len(self.parent) - 1

    def find(self, v):
        p = self.parent
        while p[v] != v:
            p[v] = p[p[v]]
            v = p[v]
        return v

    def add_word_loop(self, labels):
        """Attach a loop at base spelling `labels` (list of 0..3)."""
        cur = self.find(self.base)
        n = len(labels)
        for i, l in enumerate(labels):
            nxt = self.find(self.base) if i == n - 1 else self.new_vertex()
            self._add_edge(cur, l, nxt)
            cur = nxt

    def _add_edge(self, u, l, v):
        u, v = self.find(u), self.find(v)
        # record both directions; conflicts resolved in fold()
        self._set(u, l, v)
        self._set(v, INV[l], u)

    def _set(self, u, l, v):
        a = self.adj[u]
        if l in a:
            w = self.find(a[l])
            if w != v:
                self.pending.append((w, v))
            # keep existing; merge will unify
        else:
            a[l] = v

    def fold(self):
        # iterate: find any vertex with two same-label edges -> merge targets
        self.pending = deque()
        # initial scan: adj dicts can't hold duplicates, but merges create them via merge step
        changed = True
        while changed:
            changed = False
            for u in range(len(self.adj)):
                if self.parent[u] != u:
                    continue
                a = self.adj[u]
                for l in list(a.keys()):
                    a[l] = self.find(a[l])
            # duplicates within a dict are impossible; the real work happens in merge()
            while self.pending:
                x, y = self.pending.popleft()
                self.merge(x, y)
                changed = True

    def merge(self, x, y):
        x, y = self.find(x), self.find(y)
        if x == y:
            return
        # merge smaller adjacency into larger
        if len(self.adj[x]) < len(self.adj[y]):
            x, y = y, x
        self.parent[y] = x
        ax, ay = self.adj[x], self.adj[y]
        for l, w in ay.items():
            w = self.find(w)
            if l in ax:
                w2 = self.find(ax[l])
                if w != w2:
                    self.pending.append((w, w2))
            else:
                ax[l] = w
        self.adj[y] = {}

    def build(self, words):
        self.pending = deque()
        for labels in words:
            self.add_word_loop(labels)
        # full fold to fixpoint
        while True:
            # normalize adjacency
            for u in range(len(self.adj)):
                if self.parent[u] == u:
                    a = self.adj[u]
                    for l in list(a.keys()):
                        a[l] = self.find(a[l])
            if not self.pending:
                break
            while self.pending:
                x, y = self.pending.popleft()
                self.merge(x, y)

    def accepts(self, labels):
        v = self.find(self.base)
        for l in labels:
            a = self.adj[v]
            if l not in a:
                return False
            v = self.find(a[l])
        return v == self.find(self.base)


def word_to_labels(w):
    return [C2L[c] for c in w]

def labels_to_5th_power(root):
    return root * 5

def gen_reduced_words(maxlen):
    """All reduced words over labels 0..3, length 1..maxlen, one per {w, w^-1} pair,
    skipping proper powers (d = f^k => d^5 in <f^5>)."""
    out = []
    def inv_labels(ls):
        return [INV[l] for l in reversed(ls)]
    def is_proper_power(ls):
        n = len(ls)
        for p in range(1, n):
            if n % p == 0 and ls == ls[:p] * (n // p):
                return True
        return False
    stack = [[l] for l in range(4)]
    while stack:
        w = stack.pop()
        iw = inv_labels(w)
        if w <= iw and not is_proper_power(w):
            out.append(w)
        if len(w) < maxlen:
            for l in range(4):
                if l != INV[w[-1]]:
                    stack.append(w + [l])
    return out

def main():
    M = int(sys.argv[1])
    targets_file = sys.argv[2]
    roots = gen_reduced_words(M)
    words = [labels_to_5th_power(r) for r in roots]
    total_edges = sum(len(w) for w in words)
    print(f"M={M} roots={len(roots)} total_path_edges={total_edges}", flush=True)
    g = Graph()
    g.build(words)
    live = sum(1 for u in range(len(g.parent)) if g.parent[u] == u)
    print(f"folded: vertices={live}", flush=True)
    for line in open(targets_file):
        line = line.strip()
        if not line or line.startswith('#'):
            continue
        name, w = (line.split(None, 1) if ' ' in line else (line[:12] + '...', line))
        r = g.accepts(word_to_labels(w.strip()))
        print(f"MEMBER={r} M={M} {name}", flush=True)

if __name__ == '__main__':
    main()

File: obsidian/Experiments/Group Theory/Burnside Group/B25/Infiniteness/data/stallings/fold.rs
// Exact membership of a word in H_M = <d^5 : d reduced over {a,b}, 1 <= |d| <= M> <= F2
// via Stallings folding. Rust port of ../fold.py, scaled with compact arrays + union-find.
//
// Labels: 0=a,1=b,2=A,3=B ; inverse label = l xor 2.
// Membership(target) == target spells a closed loop at the base of the folded core.
//
// The Stallings folded core is canonical (folding is confluent), so fold order does not
// affect membership results -- this must match fold.py at every M (cross-validation).
//
// Build:  rustc -O -C opt-level=3 -C lto=fat -C codegen-units=1 fold.rs -o fold
// Run:    ./fold <M> <targets_file>

use std::env;
use std::fs;
use std::io::{self, Write, BufWriter};

const NIL: u32 = u32::MAX;

#[inline(always)]
fn inv(l: u8) -> u8 { l ^ 2 }

fn char_to_label(c: u8) -> u8 {
    match c {
        b'a' => 0,
        b'b' => 1,
        b'A' => 2,
        b'B' => 3,
        _ => panic!("bad char {}", c as char),
    }
}

// ---- Root generation: reduced words length 1..M, one per {w,w^-1}, no proper powers ----
struct RootGen {
    m: usize,
    buf: Vec<u8>,        // current word under construction
    packed: Vec<u8>,     // all emitted roots concatenated
    offsets: Vec<u32>,   // start index in `packed` of each root; last entry = packed.len()
    total_len: u64,      // sum of |d| over emitted roots
}

impl RootGen {
    fn new(m: usize) -> Self {
        RootGen { m, buf: Vec::with_capacity(m), packed: Vec::new(), offsets: vec![0], total_len: 0 }
    }

    // w == inv(reverse(w))?  and  w <= inv(reverse(w)) lexicographically.
    // Returns true iff w <= its formal inverse.
    fn le_inverse(w: &[u8]) -> bool {
        // inv-word[i] = inv(w[n-1-i]); compare w[i] vs inv(w[n-1-i]).
        let n = w.len();
        for i in 0..n {
            let a = w[i];
            let b = inv(w[n - 1 - i]);
            if a < b { return true; }
            if a > b { return false; }
        }
        true // equal
    }

    fn is_proper_power(w: &[u8]) -> bool {
        let n = w.len();
        let mut p = 1;
        while p < n {
            if n % p == 0 {
                let mut ok = true;
                for i in 0..n {
                    if w[i] != w[i % p] { ok = false; break; }
                }
                if ok { return true; }
            }
            p += 1;
        }
        false
    }

    fn emit(&mut self) {
        let w = &self.buf;
        if Self::le_inverse(w) && !Self::is_proper_power(w) {
            self.packed.extend_from_slice(w);
            self.offsets.push(self.packed.len() as u32);
            self.total_len += w.len() as u64;
        }
    }

    fn recurse(&mut self) {
        self.emit();
        if self.buf.len() >= self.m { return; }
        let last = *self.buf.last().unwrap();
        let forbidden = inv(last);
        for l in 0u8..4 {
            if l != forbidden {
                self.buf.push(l);
                self.recurse();
                self.buf.pop();
            }
        }
    }

    fn generate(&mut self) {
        for l in 0u8..4 {
            self.buf.push(l);
            self.recurse();
            self.buf.pop();
        }
    }
}

// ---- Folding graph: compact union-find + per-vertex [u32;4] adjacency ----
struct Graph {
    parent: Vec<u32>,
    rank: Vec<u8>,          // union-by-rank (canonical result; fold is confluent)
    adj: Vec<[u32; 4]>,
    stack: Vec<(u32, u32)>, // pending merges
    n: u32,                 // number of allocated vertices
}

impl Graph {
    fn with_capacity(v0: usize) -> Self {
        let mut g = Graph {
            parent: Vec::with_capacity(v0),
            rank: Vec::with_capacity(v0),
            adj: Vec::with_capacity(v0),
            stack: Vec::new(),
            n: 0,
        };
        g.new_vertex(); // base = 0
        g
    }

    #[inline]
    fn new_vertex(&mut self) -> u32 {
        let id = self.n;
        self.parent.push(id);
        self.rank.push(0);
        self.adj.push([NIL; 4]);
        self.n += 1;
        id
    }

    #[inline]
    fn find(&mut self, mut v: u32) -> u32 {
        while self.parent[v as usize] != v {
            let gp = self.parent[self.parent[v as usize] as usize];
            self.parent[v as usize] = gp; // path halving
            v = gp;
        }
        v
    }

    // Set adj[u][l]=v; on conflict (slot occupied by w!=v) queue a merge, keep existing.
    #[inline]
    fn set_edge(&mut self, u: u32, l: u8, v: u32) {
        let cur = self.adj[u as usize][l as usize];
        if cur == NIL {
            self.adj[u as usize][l as usize] = v;
        } else {
            let w = self.find(cur);
            if w != v {
                self.stack.push((w, v));
            }
        }
    }

    // Add an undirected labeled edge (both directions), fold-aware.
    #[inline]
    fn add_edge(&mut self, u: u32, l: u8, v: u32) {
        self.set_edge(u, l, v);
        self.set_edge(v, inv(l), u);
    }

    // Drain the merge worklist to fixpoint.
    fn fold(&mut self) {
        while let Some((a, b)) = self.stack.pop() {
            let mut x = self.find(a);
            let mut y = self.find(b);
            if x == y { continue; }
            if self.rank[x as usize] < self.rank[y as usize] {
                std::mem::swap(&mut x, &mut y);
            }
            self.parent[y as usize] = x;
            if self.rank[x as usize] == self.rank[y as usize] {
                self.rank[x as usize] += 1;
            }
            for l in 0..4 {
                let ty_raw = self.adj[y as usize][l];
                if ty_raw == NIL { continue; }
                self.adj[y as usize][l] = NIL;
                let ty = self.find(ty_raw);
                let tx_raw = self.adj[x as usize][l];
                if tx_raw == NIL {
                    self.adj[x as usize][l] = ty;
                } else {
                    let tx = self.find(tx_raw);
                    self.adj[x as usize][l] = tx;
                    if tx != ty {
                        self.stack.push((tx, ty));
                    }
                }
            }
        }
    }

    fn accepts(&mut self, labels: &[u8]) -> bool {
        let base = self.find(0);
        let mut v = base;
        for &l in labels {
            let t = self.adj[v as usize][l as usize];
            if t == NIL { return false; }
            v = self.find(t);
        }
        v == self.find(0)
    }

    fn live_count(&self) -> usize {
        (0..self.n as usize).filter(|&i| self.parent[i] as usize == i).count()
    }
}

fn main() {
    let args: Vec<String> = env::args().collect();
    if args.len() < 3 {
        eprintln!("usage: {} <M> <targets_file>", args[0]);
        std::process::exit(2);
    }
    let m: usize = args[1].parse().expect("M must be an integer");
    let targets_file = &args[2];

    let stdout = io::stdout();
    let mut out = BufWriter::new(stdout.lock());

    // --- generate roots ---
    let mut rg = RootGen::new(m);
    rg.generate();
    let n_roots = rg.offsets.len() - 1;
    let total_path_edges: u64 = rg.total_len * 5;
    // V0 = 1 (base) + sum over roots (5*|d| - 1)
    let v0: usize = 1 + (total_path_edges as usize - n_roots);
    writeln!(out, "M={} roots={} total_path_edges={}", m, n_roots, total_path_edges).unwrap();
    out.flush().unwrap();

    // --- build wedge of d^5 loops ---
    let mut g = Graph::with_capacity(v0);
    for r in 0..n_roots {
        let s = rg.offsets[r] as usize;
        let e = rg.offsets[r + 1] as usize;
        let root = &rg.packed[s..e];
        let l5 = root.len() * 5;
        // path base -> v1 -> ... -> v_{l5-1} -> base, spelling root repeated 5x
        let mut cur = g.find(0);
        for i in 0..l5 {
            let lab = root[i % root.len()];
            let nxt = if i == l5 - 1 { g.find(0) } else { g.new_vertex() };
            g.add_edge(cur, lab, nxt);
            cur = nxt;
        }
    }

    // --- fold ---
    g.fold();
    writeln!(out, "folded: vertices={}", g.live_count()).unwrap();
    out.flush().unwrap();

    // --- targets ---
    let content = fs::read_to_string(targets_file).expect("cannot read targets file");
    for line in content.lines() {
        let line = line.trim();
        if line.is_empty() || line.starts_with('#') { continue; }
        let (name, word) = match line.find(char::is_whitespace) {
            Some(idx) => {
                let name = &line[..idx];
                let word = line[idx..].trim();
                (name.to_string(), word.to_string())
            }
            None => {
                let name = if line.len() > 12 { format!("{}...", &line[..12]) } else { line.to_string() };
                (name, line.to_string())
            }
        };
        let labels: Vec<u8> = word.bytes().map(char_to_label).collect();
        let r = g.accepts(&labels);
        writeln!(out, "MEMBER={} M={} {}", if r { "True" } else { "False" }, m, name).unwrap();
    }
    out.flush().unwrap();
}
